crates/red_knot/src/types.rs

```diff
@@ -255,19 +262,90 @@ pub struct FunctionTypeId {
     func_id: ModuleFunctionTypeId,
 }
 
+impl FunctionTypeId {
+    pub(crate) fn function(self, db: &dyn SemanticDb) -> QueryResult<FunctionTypeRef> {
+        Ok(db.type_store()?.get_function(self))
+    }
+
+    pub(crate) fn name(self, db: &dyn SemanticDb) -> QueryResult<Name> {
+        Ok(self.function(db)?.name().into())
+    }
+
+    pub(crate) fn file(self) -> FileId {
+        self.file_id
+    }
+
+    pub(crate) fn symbol(self, db: &dyn SemanticDb) -> QueryResult<SymbolId> {
+        let FunctionType { symbol_id, .. } = *db.type_store()?.get_function(self);
+        Ok(symbol_id)
+    }
+
+    pub(crate) fn get_containing_class(
+        self,
+        db: &dyn SemanticDb,
+    ) -> QueryResult<Option<ClassTypeId>> {
+        let table = db.symbol_table(self.file_id)?;
+        let FunctionType { symbol_id, .. } = *db.type_store()?.get_function(self);
+        let scope_id = symbol_id.symbol(&table).scope_id();
+        let scope = scope_id.scope(&table);
+        match scope.kind() {
+            ScopeKind::Class => {
+                if let Some(def) = scope.definition() {
+                    if let Some(symbol_id) = scope.defining_symbol() {
+                        if let Type::Class(class) =
+                            db.infer_definition_type(self.file_id, symbol_id, def)?
+                        {
+                            Ok(Some(class))
+                        } else {
+                            Ok(None)
+                        }
+                    } else {
+                        Ok(None)
+                    }
+                } else {
+                    Ok(None)
+                }
+            }
+            _ => Ok(None),
+        }
```

Comment:
Code-style nit: I'd be inclined to "unspool" this a bit:

```suggestion
        if !matches!(scope.kind(), ScopeKind::Class) {
            return Ok(None);
        }
        let Some(def) = scope.definition() else {
            return Ok(None);
        };
        let Some(symbol_id) = scope.defining_symbol() else {
            return Ok(None);
        };
	let Type::Class(class) = db.infer_definition_type(self.file_id, symbol_id, def)? else {
	    return Ok(None);
	};
```